### PR TITLE
Possibility to duplicate a dataset from the dashboard view

### LINKED ISSUE #3423: Add Duplicate to the dashboard options when having a dataset selected
It should duplicate the dataset with the same name+_copy and appear in the list of tables.

### LINKED ISSUE #4353: Add duplicate as an option for datasets shared with you
- In the dashboard.
- In the dataset view

### PR DESCRIPTION
**Pull request by xavijam**
- [x] Duplicating one of your datasets (Fixes #3423).

screen shot 2015-08-19 at 13 44 43

--- 

- [x] Duplicating one of the datasets shared with you within your organization (Fixes #4353).

screen shot 2015-08-19 at 13 44 27

---

- Duplicating one of the datasets shared with you within your organization in the editor (option already available).

### CODE REVIEW

File: lib/assets/javascripts/cartodb/dashboard/filters_view.js [hunk 1]
```diff
@@ -290,6 +292,47 @@ module.exports = cdb.core.View.extend({
...
+      cdb.god.trigger('importByUploadData', {
+        type: 'duplicate',
+        table_name: table.getUnqualifiedName() + '_copy',
+        value: tableName,
+        create_vis: false
+      });
+    }
+  },
+
+  _createDatasetFromSQL: function(e) {
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: Import from query is not used for the moment, but ready when it is needed.

File: lib/assets/javascripts/cartodb/dashboard/filters_view.js [hunk 2]
```diff
@@ -290,6 +292,47 @@ module.exports = cdb.core.View.extend({
...
+
+    var selectedDatasets = this._selectedItems();
+
+    if (selectedDatasets.length === 1) {
+      var m = selectedDatasets[0];
+      var table = m.tableMetadata();
+      var owner = table.permission && table.permission.owner.get('username');
+      var tableName = table.getUnqualifiedName();
+      var query = encodeURIComponent("SELECT * FROM " + tableName);
+      var sqlApiUrl = cdb.config.getSqlApiUrl().replace(this.user.get('username'), owner);
```
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: Question: What would it be the best way to generate the SQL API url from other user? (not the current one).

**Comment by xavijam** [comment]
Could you review this @viddo?
Also, I feel you can take advantage of this in your "cleaning" PR, because duplicate and sql export are available within new import model.

File: lib/assets/javascripts/cartodb/dashboard/filters_view.js (hunk 2 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: Why would you want to generate a SQL API URL for another user? Shouldn't they always be done from the POV of the current user?

Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: Example, we would like to duplicate a dataset you liked. The only way to do it is using the API SQL, and the url should be:
```
http://thatuser.cartodb.com/api/vX/sql?q=SELECT * FROM anytable
```

**Comment by viddo** [comment]
The "cleaning" PR was for removing old modal code, so I don't consider this to be in that scope. 

Using this new code would be an "enhancement" (in terms of code maintenance at least, for the end-user there's no difference), similar to what is planned in #5076, consolidating models/logic to have one "proper" way of doing a particular action. Thus, if you think that's the case I'd suggest to create a separate issue instead.

**Comment by xavijam** [comment]
Yep, you are right @viddo, that should be another issue. Let me create it (should I assign it to you?).

File: lib/assets/javascripts/cartodb/dashboard/filters_view.js (hunk 1 above)
Comment (viddo) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: IMHO, if this code isn't used then I'd say remove it. Keep a gist for the future if you want to remember this particular code. That also postpones the discussion below until we actually need to solve that problem and know all the real requirements.

**Comment by viddo** [comment]
Depends on fast you'd like this to get done, I have 20-isch issues pending for upcoming milestones. ;)

File: lib/assets/javascripts/cartodb/dashboard/filters_view.js (hunk 1 above)
Comment (xavijam) [inline lib/assets/javascripts/cartodb/dashboard/filters_view.js]: :+1:

**Pull request merged**